### PR TITLE
Three-Pane Diff and minor fixes

### PR DESCRIPTION
**Pull request by PurpleGuitar**
This change introduces a new feature, as well as several bug fixes.

Feature: The diff page provided by the 'revisionsfull' plugin has been modified to allow a three-way comparison between versions.  When the URL to the diff view contains three revisions and a difftype of 'rendered', the diff page will show three panes instead of two.  The new third pane has similar functionality to the existing pages, including the revision drop-down.  The scrollbars of all three panes can be locked to allow easier scrolling.

Fix: Only show chunk links in allowed areas of Dokuwiki, currently en:bible:notes:\* and en:obe:*.

Fix: Always show chunk links where they are allowed, instead of suppressing them when no status tags are present.

Fix: Correct possible null reference when calculating where status changes between revisions.

[Review on Reviewable](https://reviewable.io/reviews/unfoldingword-dev/door43/402)

### CODE REVIEW

**Comment by phillip-hopper** [comment]
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

_[lib/plugins/chunkprogress/action.php, line 169 [r1]](https://reviewable.io:443/reviews/unfoldingword-dev/door43/402#-K56PkfveFC5KJ0uhImX:-K56PkfveFC5KJ0uhImY:927143768) ([raw file](https://github.com/unfoldingword-dev/door43/blob/df5d8f7e12074760f298cb4911db0860b47e7bd3/lib/plugins/chunkprogress/action.php#L169)):_
It looks like there is inconsistent indenting here.

---

_[lib/plugins/chunkprogress/action.php, line 176 [r1]](https://reviewable.io:443/reviews/unfoldingword-dev/door43/402#-K56Puhq7DG-hhQ-m5QO:-K56Puhq7DG-hhQ-m5QP:630651461) ([raw file](https://github.com/unfoldingword-dev/door43/blob/df5d8f7e12074760f298cb4911db0860b47e7bd3/lib/plugins/chunkprogress/action.php#L176)):_
Inconsistent indenting here also.

---

_[lib/plugins/chunkprogress/action.php, line 183 [r1]](https://reviewable.io:443/reviews/unfoldingword-dev/door43/402#-K56Pz1-zFyqwAbYPk8j:-K56Pz1-zFyqwAbYPk8k:630651461) ([raw file](https://github.com/unfoldingword-dev/door43/blob/df5d8f7e12074760f298cb4911db0860b47e7bd3/lib/plugins/chunkprogress/action.php#L183)):_
Inconsistent indenting here also.

---

_[lib/plugins/revisionsfull/action.php, line 506 [r1]](https://reviewable.io:443/reviews/unfoldingword-dev/door43/402#-K56RK2-EnVvbNLOP7b-:-K56RK2-EnVvbNLOP7b0:-709063006) ([raw file](https://github.com/unfoldingword-dev/door43/blob/df5d8f7e12074760f298cb4911db0860b47e7bd3/lib/plugins/revisionsfull/action.php#L506)):_
33 x 3 = 99%.  It might be a good idea to make the total 100%.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/unfoldingword-dev/door43/402)_

**Comment by PurpleGuitar** [comment]
Thanks for your comments Phil; I'll fix those issues.

**Comment by PurpleGuitar** [comment]
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

_[lib/plugins/chunkprogress/action.php, line 169 [r1]](https://reviewable.io:443/reviews/unfoldingword-dev/door43/402#-K56PkfveFC5KJ0uhImX:-K5BRb1_38lBbuVtlZRy:816682731) ([raw file](https://github.com/unfoldingword-dev/door43/blob/df5d8f7e12074760f298cb4911db0860b47e7bd3/lib/plugins/chunkprogress/action.php#L169)):_
Fixed: I unwrapped the line.

---

_[lib/plugins/chunkprogress/action.php, line 176 [r1]](https://reviewable.io:443/reviews/unfoldingword-dev/door43/402#-K56Puhq7DG-hhQ-m5QO:-K5BRrWgH8ya9eIAi_Ib:816682731) ([raw file](https://github.com/unfoldingword-dev/door43/blob/df5d8f7e12074760f298cb4911db0860b47e7bd3/lib/plugins/chunkprogress/action.php#L176)):_
Fixed: I unwrapped the line.

---

_[lib/plugins/chunkprogress/action.php, line 183 [r1]](https://reviewable.io:443/reviews/unfoldingword-dev/door43/402#-K56Pz1-zFyqwAbYPk8j:-K5BRsP3zsPSszNCYKA2:816682731) ([raw file](https://github.com/unfoldingword-dev/door43/blob/df5d8f7e12074760f298cb4911db0860b47e7bd3/lib/plugins/chunkprogress/action.php#L183)):_
Fixed: I unwrapped the line.

---

_[lib/plugins/revisionsfull/action.php, line 506 [r1]](https://reviewable.io:443/reviews/unfoldingword-dev/door43/402#-K56RK2-EnVvbNLOP7b-:-K5BS1PUUtWkpOlTO6Op:-1154985047) ([raw file](https://github.com/unfoldingword-dev/door43/blob/df5d8f7e12074760f298cb4911db0860b47e7bd3/lib/plugins/revisionsfull/action.php#L506)):_
I kept the panes the same size so that the text would wrap identically.  If the panes are even slightly different sizes, they will wrap differently and look as though the text has changed, even though it hasn't.

I tried some absolute measures (e.g. "10em"), but it skewed the panes.

Are you OK with keeping them at 33%?

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/unfoldingword-dev/door43/402)_

**Comment by phillip-hopper** [comment]
Reviewed 3 of 4 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/unfoldingword-dev/door43/402)_

File: lib/plugins/chunkprogress/action.php [hunk 1]
```diff
@@ -83,15 +83,34 @@ private function should_process_page() {
...
+
+        // Whitelist -- reject pages that don't match regex
+        $whitelisted = false;
+        $whitelist_namespaces = array(
+            "/en:bible:notes:.*/",
+            "/en:obe:.*/"
+        );
+        foreach ($whitelist_namespaces as $whitelist_namespace) {
+            if (preg_match($whitelist_namespace, $ID)) {
+                $whitelisted = true;
```
Comment (PurpleGuitar) [inline lib/plugins/chunkprogress/action.php:99]: Break after being whitelisted -- thanks Joe!

**Pull request merged**